### PR TITLE
Replacing NotNullAttribute with thrown exceptions

### PR DESCRIPTION
**Pull request by pranavkm**
Also adding `warningAsErrors: true` since I'm already touching project.jsons

### CODE REVIEW

File: src/Microsoft.AspNet.Http.Abstractions/Extensions/MapWhenExtensions.cs [hunk 1]
```diff
@@ -2,10 +2,9 @@
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
 using System;
-using System.Threading.Tasks;
 using Microsoft.AspNet.Http;
 using Microsoft.AspNet.Builder.Extensions;
-using Microsoft.Framework.Internal;
+
```
Comment (Tratcher) [inline src/Microsoft.AspNet.Http.Abstractions/Extensions/MapWhenExtensions.cs]: nit: spacing.

File: src/Microsoft.AspNet.Http.Abstractions/Extensions/MapWhenOptions.cs [hunk 2]
```diff
@@ -13,14 +11,26 @@ namespace Microsoft.AspNet.Builder.Extensions
...
-            get;
-            [param: NotNull]
-            set;
+            get
+            {
+                return _predicate;
+            }
+            set
+            {
+                if (value == null)
```
Comment (Tratcher) [inline src/Microsoft.AspNet.Http.Abstractions/Extensions/MapWhenOptions.cs]: Keep the Options simple and move this validation into the middleware constructor.

File: src/Microsoft.AspNet.Http.Features/Authentication/ChallengeContext.cs [hunk 3]
```diff
@@ -3,18 +3,26 @@
...
-using Microsoft.Framework.Internal;
 
 namespace Microsoft.AspNet.Http.Features.Authentication
 {
     public class ChallengeContext
     {
-        public ChallengeContext([NotNull] string authenticationScheme) : this(authenticationScheme, properties: null, behavior: ChallengeBehavior.Automatic)
+        public ChallengeContext(string authenticationScheme) : this(authenticationScheme, properties: null, behavior: ChallengeBehavior.Automatic)
         {
+            if (authenticationScheme == null)
```
Comment (Tratcher) [inline src/Microsoft.AspNet.Http.Features/Authentication/ChallengeContext.cs]: redundant, handled in chained constructor.

File: src/Microsoft.AspNet.Http/Features/RequestCookiesFeature.cs [hunk 4]
```diff
@@ -19,18 +18,32 @@ public class RequestCookiesFeature : IRequestCookiesFeature
         private StringValues _original;
         private IReadableStringCollection _parsedValues;
 
-        public RequestCookiesFeature([NotNull] IDictionary<string, StringValues> cookies)
+        public RequestCookiesFeature(IDictionary<string, StringValues> cookies)
             : this(new ReadableStringCollection(cookies))
         {
+            if (cookies == null)
```
Comment (Tratcher) [inline src/Microsoft.AspNet.Http/Features/RequestCookiesFeature.cs]: redundant, chaining.

File: src/Microsoft.AspNet.Http/FormCollection.cs [hunk 5]
```diff
@@ -12,14 +12,28 @@ namespace Microsoft.AspNet.Http.Internal
     /// </summary>
     public class FormCollection : ReadableStringCollection, IFormCollection
     {
-        public FormCollection([NotNull] IDictionary<string, StringValues> store)
+        public FormCollection(IDictionary<string, StringValues> store)
             : this(store, new FormFileCollection())
         {
+            if (store == null)
```
Comment (Tratcher) [inline src/Microsoft.AspNet.Http/FormCollection.cs]: chaining

File: src/Microsoft.AspNet.Http/FormCollection.cs [hunk 6]
```diff
@@ -12,14 +12,28 @@ namespace Microsoft.AspNet.Http.Internal
...
+            {
+                throw new ArgumentNullException(nameof(store));
+            }
         }
 
-        public FormCollection([NotNull] IDictionary<string, StringValues> store, [NotNull] IFormFileCollection files)
+        public FormCollection(IDictionary<string, StringValues> store, IFormFileCollection files)
             : base(store)
         {
+            if (store == null)
```
Comment (Tratcher) [inline src/Microsoft.AspNet.Http/FormCollection.cs]: chaining

File: src/Microsoft.AspNet.WebUtilities/MultipartReader.cs [hunk 7]
```diff
@@ -21,13 +20,31 @@ public class MultipartReader
         private readonly string _boundary;
         private MultipartReaderStream _currentStream;
 
-        public MultipartReader([NotNull] string boundary, [NotNull] Stream stream)
+        public MultipartReader(string boundary, Stream stream)
             : this(boundary, stream, DefaultBufferSize)
         {
+            if (boundary == null)
```
Comment (Tratcher) [inline src/Microsoft.AspNet.WebUtilities/MultipartReader.cs]: chaining

File: src/Microsoft.Net.Http.Headers/SetCookieHeaderValue.cs [hunk 8]
```diff
@@ -33,13 +32,27 @@ private SetCookieHeaderValue()
             // Used by the parser to create a new instance of this type.
         }
 
-        public SetCookieHeaderValue([NotNull] string name)
+        public SetCookieHeaderValue(string name)
             : this(name, string.Empty)
         {
+            if (name == null)
```
Comment (Tratcher) [inline src/Microsoft.Net.Http.Headers/SetCookieHeaderValue.cs]: chaining

**Comment by Tratcher** [comment]
Minor comments, then :shipit:

File: src/Microsoft.AspNet.Http.Abstractions/Extensions/MapWhenOptions.cs (hunk 2 above)
Comment (pranavkm) [inline src/Microsoft.AspNet.Http.Abstractions/Extensions/MapWhenOptions.cs]: We do have follow this pattern in Mvc. Plus it has the added advantage of throwing when you set it as opposed to when we create the middleware. I'll leave this as is.

**Pull request closed without merging**